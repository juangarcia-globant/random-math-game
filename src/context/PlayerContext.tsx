import { useState, createContext, useContext, ReactNode } from 'react';

type PlayerProviderProps = {children: ReactNode};

const PlayerStateContext = createContext<
{username: string, setUsername: React.Dispatch<React.SetStateAction<string>>} | undefined
>(undefined);

function PlayerProvider ({ children }: PlayerProviderProps) {
  const [username, setUsername] = useState<string>('');

  return (
    <PlayerStateContext.Provider value={{ username, setUsername }}>
      { children }
    </PlayerStateContext.Provider>
  );
}

function usePlayerContext () {
  const context = useContext(PlayerStateContext);

  if (context === undefined) {
    throw new Error('usePlayerContext must be inside a PlayerProvider');
  }

  return context;
}

export { PlayerProvider, usePlayerContext };