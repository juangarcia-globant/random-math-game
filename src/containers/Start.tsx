import { useState } from 'react';
// Components
import { Input } from '../components/Input';
import { Button } from '../components/Button';
// Context
import { usePlayerContext } from '../context/PlayerContext';

type StartProps = {
  onClick: () => void,
};

export const Start: React.FunctionComponent<StartProps> = ({
  onClick,
}) => {
  const [error, setError] = useState({
    message: '',
    hasError: false,
  });

  const { username, setUsername } = usePlayerContext();

  const setNameOfPlayer = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;

    if (error.hasError && value) {
      setError({ message: '', hasError: false });
    }
    setUsername(value)
  }

  const handleOnKeyDown = (evt) => {
    if (evt.key === 'Enter'){
      handleOnClick();
    }
  }

  const handleOnClick = () => {
    if (!username) {
      setError({
        message: 'Please insert a username.',
        hasError: true,
      });
      return;
    };
    onClick();
  }

  return (
    <>
      <p>
        Insert your username <i className="fas fa-graduation-cap" />
      </p>
      <Input
        type="text"
        onKeyDown={handleOnKeyDown}
        onInputChange={setNameOfPlayer}
        error={error.message}
        hasError={error.hasError}
      />
      <p> Then press to start the game </p>
      <Button text="Start" onClick={handleOnClick} />
    </>
  )
}
