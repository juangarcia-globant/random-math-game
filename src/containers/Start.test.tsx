import { render, screen } from "@testing-library/react"
import { PlayerProvider } from "../context/PlayerContext";
import { Start } from "./Start";

describe("<Start />", () => {
  test("Matches the snapshot", () => {
    const { asFragment } = render(
      <PlayerProvider>
        <Start onClick={() => {}} />
      </PlayerProvider>
    );
    expect(asFragment()).toMatchSnapshot()
  });

  test("Start container should be rendered", () => {
    render(
      <PlayerProvider>
        <Start onClick={() => {}} />
      </PlayerProvider>
    );
    expect(screen.getByText('Insert your username')).toBeInTheDocument();
    expect(screen.getByText('Then press to start the game')).toBeInTheDocument();
    expect(screen.getByText('Start')).toBeInTheDocument();
  });
});