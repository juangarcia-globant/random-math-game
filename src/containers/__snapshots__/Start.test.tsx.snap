// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Start /> Matches the snapshot 1`] = `
<DocumentFragment>
  <p>
    Insert your username 
    <i
      class="fas fa-graduation-cap"
    />
  </p>
  <input
    class="input input-sm"
    id="textId"
    type="text"
  />
  <p>
     Then press to start the game 
  </p>
  <button
    class="button"
  >
    Start
  </button>
</DocumentFragment>
`;
